### PR TITLE
src: lib: dialog_ui: Add async loading function

### PR DESCRIPTION
**Pull request by Andregnl**
Add async loading ui with an animated spinner.

To call the async method the caller needs to store the pid of the function call and use the method stop_loading() afterwards.

### CODE REVIEW

**Comment by rodrigosiqueira** [comment]
Nice, but can you add this spin when we access a mailing list via the 'Registered Mailing list' menu?

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:244]: Add simple documentation about this function. Also, use a camel case.

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:249]: Remove this extra space

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:248]: `char_width`

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:250]: Add `""` around `$1`

Comment (rodrigosiqueira) [reply src/lib/dialog_ui.sh:250]: Also, add a new empty line.

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:280]: Why not just use the numbers without quotes?

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:285]: Keep variable declaration together at the beginning of this function.

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:293]: Do we really need this inside the loop?

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:313]: I was wondering if signals would be a better way to handle this.

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:309]: Document.

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [inline src/lib/dialog_ui.sh:311]: Double-quotes.

File: src/lib/lore.sh
Comment (rodrigosiqueira) [inline src/lib/lore.sh:421]: Is this indentation correct?

File: src/lib/lore.sh
Comment (rodrigosiqueira) [inline src/lib/lore.sh:430]: Double quotes (check in the entire PR).

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

File: src/lib/dialog_ui.sh
Comment (davidbtadokoro) [reply src/lib/dialog_ui.sh:244]: Sorry, for intruding, but don't you mean `snake_case`?

File: src/ui/patch_hub/latest_patchsets_from_mailing_list.sh
Comment (davidbtadokoro) [inline src/ui/patch_hub/latest_patchsets_from_mailing_list.sh:19]: Double-quotes.

File: src/ui/patch_hub/lore_mailing_lists.sh
Comment (davidbtadokoro) [inline src/ui/patch_hub/lore_mailing_lists.sh:21]: Double-quotes.

File: src/ui/patch_hub/lore_mailing_lists.sh
Comment (davidbtadokoro) [inline src/ui/patch_hub/lore_mailing_lists.sh:29]: Double-quotes.

File: src/ui/patch_hub/patchset_details_and_actions.sh
Comment (davidbtadokoro) [inline src/ui/patch_hub/patchset_details_and_actions.sh:117]: Double-quotes.

File: src/lib/lore.sh
Comment (davidbtadokoro) [inline src/lib/lore.sh:461]: Maybe add a succinct comment to explain the why we need this for loop (think it's to not conflict with other background processes created outside this function, right?).

File: src/lib/dialog_ui.sh
Comment (davidbtadokoro) [inline src/lib/dialog_ui.sh:270]: @Andregnl and @fernandohjunqueira, this is about the below problem that happens when running the tests that we talked offline.

photo_2023-10-31 01 04 53

The problem is that the test case `test_handle_bookmark_action` of the test file `tests/ui/patch_hub/patchset_details_and_actions_test.sh` tests the function `handle_bookmark_action`, which calls `create_async_loading_screen_notification` underneath, and makes this dialog screen appear without failing the test.

The way to solve this is by mocking the function inside the test case (like I showed you in the past) to avoid creating a dialog screen. This mock is a little more complicated, because `handle_bookmark_action` also calls `stop_loading`, so we need to make this process continuous. Below is the diff to ease your work

```bash
diff --git a/tests/ui/patch_hub/patchset_details_and_actions_test.sh b/tests/ui/patch_hub/patchset_details_and_actions_test.sh
index 0108dd1..12afc4d 100755
--- a/tests/ui/patch_hub/patchset_details_and_actions_test.sh
+++ b/tests/ui/patch_hub/patchset_details_and_actions_test.sh
@@ -154,9 +154,11 @@ test_handle_bookmark_action()
   local output
 
   # shellcheck disable=SC2317
-  function create_loading_screen_notification()
+  function create_async_loading_screen_notification()
   {
-    return
+    while true; do
+      sleep 0.1
+    done
   }
 
   # We need to mock download_series or else the test
```

Ping me if you don't understand my explanation 😬

File: src/lib/dialog_ui.sh
Comment (davidbtadokoro) [reply src/lib/dialog_ui.sh:309]: Also, if this function is dedicated to the `create_async_loading_screen_notification` function, revise the name to something like `stop_async_loading_screen_notification`, or the like.

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hey @Andregnl and @fernandohjunqueira, and thanks for the update! This is a well-waited change for the `patch-hub` feature.

Left some comments, so please see if you can address them. Also, don't forget to address the ones @rodrigosiqueira left. Remember to favor seeing the review in the 'Commits' tab of this PR, as the comments will appear inline (and this is also the way we maintainers review).

In terms of behaviour, really liked it!!! Works as a charm 😆

File: src/lib/dialog_ui.sh
Comment (rodrigosiqueira) [reply src/lib/dialog_ui.sh:244]: you are right, snake_case.

File: src/lib/dialog_ui.sh
Comment (fernandohjunqueira) [reply src/lib/dialog_ui.sh:280]: We've just copied this style from create_loading_screen_notification function.

Should we change both?

File: src/lib/dialog_ui.sh
Comment (fernandohjunqueira) [reply src/lib/dialog_ui.sh:293]: We cannot remove it from the loop, because `cmd` must be reset on each iteration. If we put those lines outside of the while loop, the loop will keep appending `" --infobox $'${loading_message} ${spin}'"` to cmd without cleaning its last value.

Comment (davidbtadokoro) [reply src/lib/dialog_ui.sh:293]: Oh, I see it now! Think you are right, so just ignore our comments. You can mark this conversation as resolved, also.

File: src/lib/dialog_ui.sh
Comment (davidbtadokoro) [inline src/lib/dialog_ui.sh:304]: Remove these 2 lines: the debug line and the blank line.

File: src/ui/patch_hub/patchset_details_and_actions.sh
Comment (davidbtadokoro) [inline src/ui/patch_hub/patchset_details_and_actions.sh:167]: Nitpick alert: delete this extra blank line to maintain the "pattern" you used in the other files.

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hey @Andregnl and @fernandohjunqueira, thanks for the new update.

I only left two comments, a minor slip and a nitpicky one 😬. Please, see if you can address them as they are really simple.

@rodrigosiqueira, in my offline conversation with them, we were a little confused about your comment below

> I was wondering if signals would be a better way to handle this.

See if what they done, i.e., use `kill -15 "$loading_pid"` to make it explicit the signal TERM is being sent, was what you were thinking.

**Comment by Andregnl** [comment]
@davidbtadokoro We made all the corrections, waiting for your review.

**Comment by davidbtadokoro** [comment]
@Andregnl and @fernandohjunqueira, thank you so much for listening to our feedback and keep updating this PR.

Merged into the unstable! 👍 

P.S.: As I mentioned in our offline conversation, I amended the commit including a unit test for the added Dialog function.

**Pull request closed without merging**
